cardinal/server/events.go

```diff
@@ -1,32 +1,159 @@
 package server
 
 import (
-	"log"
 	"net/http"
+	"time"
+
+	"github.com/go-openapi/runtime/middleware"
 
 	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog/log"
 )
 
+func CreateEventHub() *EventHub {
+	res := EventHub{
+		websocket_connections: map[*websocket.Conn]bool{},
+		Broadcast:             make(chan []byte),
+		Register:              make(chan *websocket.Conn),
+		Unregister:            make(chan *websocket.Conn),
+		Shutdown:              make(chan bool),
+	}
+	return &res
+}
+
+type Event struct {
+	message string
+}
+
+type EventHub struct {
+	websocket_connections map[*websocket.Conn]bool
+	Broadcast             chan []byte
+	Unregister            chan *websocket.Conn
+	Register              chan *websocket.Conn
+	Shutdown              chan bool
+}
+
+func (h *EventHub) run() {
+Loop:
+	for {
+		select {
+		case conn := <-h.Register:
+			h.websocket_connections[conn] = true
+		case conn := <-h.Unregister:
+			if _, ok := h.websocket_connections[conn]; ok {
+				delete(h.websocket_connections, conn)
+				err := conn.Close()
+				if err != nil {
+					log.Logger.Error().Err(err)
+				}
+			}
+		case message := <-h.Broadcast:
+			for conn := range h.websocket_connections {
+
+				err := conn.SetWriteDeadline(time.Now().Add(3 * time.Second))
+				if err != nil {
+					go func() {
+						h.Unregister <- conn
+					}()
+					log.Logger.Error().Err(err)
+					break Loop
```

Comment:
This doesn't seem right. If ANY connection fails their SetWriteDeadline call we bail out of run entirely?

Reply:
yeah you're right, should not bail for a deadline failure. Will fix.

Reply:
Oh wait. This should be true. Yes this means something is wrong. Getting a logic error on SerWriteDeadline is akin to a 500 error. It should never error unless the underlying connection has an issue. 

I think what should be fixed is if the WriteMessage call errors out BECAUSE of a deadline. That one has to be checked. Right now it's bailing.

Reply:
Documentation says that websocket connections completely close on deadline. So that means it's dead forever. I'll set deadline to 30s.

Reply:
👍